### PR TITLE
LUCENE-9446: In boolean rewrite, remove MatchAllDocsQuery filter clauses

### PR DESCRIPTION
**Pull request by jtibshirani**
Previously, we only removed 'match all' FILTER clauses if there was at least one
MUST clause. Now they're also removed if there is another distinct FILTER clause.

This lets boolean queries like `#field:value #*:*` be written to `#field:value`.

### CODE REVIEW

File: lucene/core/src/test/org/apache/lucene/search/TestBooleanRewrites.java [hunk 1]
```diff
@@ -312,12 +312,20 @@ public void testRemoveMatchAllFilter() throws IOException {
...
         .build();
     assertEquals(expected, searcher.rewrite(bq));
+
+    bq = new BooleanQuery.Builder()
+            .add(new TermQuery(new Term("foo", "bar")), Occur.FILTER)
+            .add(new MatchAllDocsQuery(), Occur.FILTER)
+            .build();
+    expected = new BoostQuery(new ConstantScoreQuery(
+            new TermQuery(new Term("foo", "bar"))), 0.0f);
+    assertEquals(expected, searcher.rewrite(bq));
```
Comment (jpountz) [inline lucene/core/src/test/org/apache/lucene/search/TestBooleanRewrites.java:328]: maybe also add a test for the case when there are two `MatchAllDocsQuery` queries as filters with no `MUST` clauses?

Comment (jtibshirani) [reply lucene/core/src/test/org/apache/lucene/search/TestBooleanRewrites.java:328]: 👍

**Comment by jtibshirani** [comment]
Thanks for the review, I added another test and a changelog entry.

**Comment by jpountz** [comment]
FYI I backported to 8.7 and moved the CHANGES entry from 9.0 to 8.7.

**Pull request merged**